clients/src/main/java/org/apache/kafka/clients/consumer/AbstractPartitionAssignor.java

```diff
@@ -0,0 +1,331 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.clients.consumer;
+
+import org.apache.kafka.common.Cluster;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.protocol.types.ArrayOf;
+import org.apache.kafka.common.protocol.types.Field;
+import org.apache.kafka.common.protocol.types.GenericType;
+import org.apache.kafka.common.protocol.types.Schema;
+import org.apache.kafka.common.protocol.types.SchemaException;
+import org.apache.kafka.common.protocol.types.Struct;
+import org.apache.kafka.common.protocol.types.Type;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Abstract assignor implementation used by default assignors shipped with Kafka. The embedded
+ * format of assignments and subscriptions are as follows:
+ *
+ * <pre>
+ * Subscription => Version Topics
+ *   Version       => int16
+ *   Topics        => [String]
+ *
+ * Assignment => Version TopicPartitions
+ *   Version         => int16
+ *   TopicPartitions => [Topic Partitions]
+ *     Topic         => String
+ *     Partitions    => [int32]
+ * </pre>
+ *
+ * This assignor assumes that any changes to these formats will be compatible with this format.
+ * In other words, newer versions can be parsed using the old format. This generally means that
+ * fields can be added to the end of the structures, but no existing fields can be removed or reordered.
+ */
+public abstract class AbstractPartitionAssignor implements PartitionAssignor<AbstractPartitionAssignor.Subscription, AbstractPartitionAssignor.Assignment> {
+    private static final Logger log = LoggerFactory.getLogger(AbstractPartitionAssignor.class);
+
+    public static final String VERSION_KEY_NAME = "version";
```

Comment:
Ok, so here's a summary of the changes I'm suggesting:

1\. Change consumer metadata schema to the following:
```
Subscription => Version Topics UserData
  Version => Int16
  Topics => [String]
  UserData => Bytes
```
Assignment stays as it currently is (are there any cases where an assignor implementation would need to propagate additional information in the assignment)?

2\. Migrate the subscription/assignment schemas currently in AbstractPartitionAssignor back into ConsumerProtocol.

3\. Change the PartitionAssignor interface to look something like this:
```java
interface PartitionAssignor {
  Subscription subscription(List<String> topics);
  Map<String, List<TopicPartition>> assign(Cluster metadata, Map<String, Subscription> subscriptions);

  class Subscription {
    ByteBuffer userData;
    List<String> topics;
  }
}
```

4\. Add ProtocolType field to JoinGroup request:
```
JoinGroupRequest => GroupId SessionTimeout MemberId ProtocolType Protocols
  GroupId => String
  SessionTimeout => int32
  MemberId => String
  ProtocolType => String
  Protocols => [Protocol MemberMetadata]
    Protocol => String
    MemberMetadata => Bytes
```
For the consumer, the protocol type will be "consumer" and this will allow tools to use ConsumerProtocol to parse subscription/assignment metadata (assuming we will eventually expose it in DescribeGroup or something similar).

@guozhangwang @ewencp What do you think?